linked_list/linked_list.py

```diff
@@ -86,32 +156,61 @@ def visit(self):
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def reverse(self):
-        pass
-  
+        previous = None
+        current = self.head
+        next_node = None
+
+        while current:
+            next_node = current.next
+            current.next = previous
+            previous = current
+            current = next_node
+        self.head = previous
+
+
     ## Advanced/ Exercises
     # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_middle_value(self):
-        pass
+        list_length = self.length()
+        middle_index = list_length // 2
+        return self.get_at_index(middle_index)
+
+
 
     # find the nth node from the end and return its value
     # assume indexing starts at 0 while counting to n
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_nth_from_end(self, n):
-        pass
+        list_length = self.length()
+        index = list_length - n - 1
+        if index < 0:
+            return None
+        return self.get_at_index(index)
+
 
     # checks if the linked list has a cycle. A cycle exists if any node in the
     # linked list links to a node already visited.
     # returns true if a cycle is found, false otherwise.
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(?)
+    # Space Complexity: O(?)
     def has_cycle(self):

```

Comment:
👍  Time/space complexity?

Good use of a set.

Reply:
Whoops! Time Complexity: O(n)? And Space Complexity: O(n)?